p2p/chain.py

```diff
@@ -437,19 +465,60 @@ def __init__(self,
 
         return dict((part.unique_key, part.part) for part in parts)
 
+    def _get_throughput_stats(self, peer: BasePeer, body_part: BlockPartEnum) -> ThroughputTracker:
+        return self._throughput_stats[body_part][peer]
+
+    def deregister_peer(self, peer: BasePeer) -> None:
+        # Delete all in-progress stats for the given peer. Otherwise, if a peer disconnects before
+        # sending a block part, stats would fail when starting to track throughput the 2nd time.
+        # Also, we don't want a perpetually growing dict of stats.
+        for part in BlockPartEnum:
+            del self._throughput_stats[part][peer]
+
     def _request_block_parts(
             self,
             target_td: int,
             headers: List[BlockHeader],
-            request_func: Callable[[ETHPeer, List[BlockHeader]], None]) -> int:
+            request_func: Callable[[ETHPeer, List[BlockHeader]], None],
+            block_part: BlockPartEnum,
+    ) -> int:
+        """
+        :return: how many requests were made (one request per peer)
+        """
         peers = self.peer_pool.get_peers(target_td)
         if not peers:
             raise NoEligiblePeers()
-        length = math.ceil(len(headers) / len(peers))
-        batches = list(partition_all(length, headers))
+        batches = self._select_headers_by_peer(headers, peers, block_part)
         for peer, batch in zip(peers, batches):
-            request_func(cast(ETHPeer, peer), batch)
-        return len(batches)
+            if len(batch):
+                self._get_throughput_stats(peer, block_part).begin_work()
+                request_func(cast(ETHPeer, peer), batch)
+        return len([batch for batch in batches if len(batch)])
+
+    def _select_headers_by_peer(
+            self,
+            headers: List[BlockHeader],
+            peers: List[BasePeer],
+            block_part: BlockPartEnum,
+    ) -> List[List[BlockHeader]]:
+        speeds = [self._get_throughput_stats(peer, block_part).get_throughput() for peer in peers]
+        speed_sum = sum(speeds)
+        fractional_speeds = [speed / speed_sum for speed in speeds]
+
+        # request headers proportionally, so faster peers are asked for more headers than slow peers
```

Comment:
The peers will actually be asked for body parts for certain headers, and not for headers themselves, right? The comment kind of suggests the latter